File: Project Academy/06. LiveData/java/app/src/test/java/com/dicoding/academies/data/FakeAcademyRepository.java
package com.dicoding.academies.data;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.dicoding.academies.data.source.local.entity.ContentEntity;
import com.dicoding.academies.data.source.local.entity.CourseEntity;
import com.dicoding.academies.data.source.local.entity.ModuleEntity;
import com.dicoding.academies.data.source.remote.RemoteDataSource;
import com.dicoding.academies.data.source.remote.response.CourseResponse;
import com.dicoding.academies.data.source.remote.response.ModuleResponse;

import java.util.ArrayList;
import java.util.List;

public class FakeAcademyRepository implements AcademyDataSource {

    private volatile static FakeAcademyRepository INSTANCE = null;

    private final RemoteDataSource remoteDataSource;

    FakeAcademyRepository(@NonNull RemoteDataSource remoteDataSource) {
        this.remoteDataSource = remoteDataSource;
    }

    public static FakeAcademyRepository getInstance(RemoteDataSource remoteData) {
        if (INSTANCE == null) {
            synchronized (FakeAcademyRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new FakeAcademyRepository(remoteData);
                }
            }
        }
        return INSTANCE;
    }

    @Override
    public ArrayList<CourseEntity> getAllCourses() {
        List<CourseResponse> courseResponses = remoteDataSource.getAllCourses();
        ArrayList<CourseEntity> courseList = new ArrayList<>();
        for (CourseResponse response : courseResponses) {
            CourseEntity course = new CourseEntity(response.getId(),
                    response.getTitle(),
                    response.getDescription(),
                    response.getDate(),
                    false,
                    response.getImagePath());

            courseList.add(course);
        }
        return courseList;
    }

    @Override
    public ArrayList<CourseEntity> getBookmarkedCourses() {
        List<CourseResponse> courseResponses = remoteDataSource.getAllCourses();
        ArrayList<CourseEntity> courseList = new ArrayList<>();
        for (CourseResponse response : courseResponses) {
            CourseEntity course = new CourseEntity(response.getId(),
                    response.getTitle(),
                    response.getDescription(),
                    response.getDate(),
                    false,
                    response.getImagePath());
            courseList.add(course);
        }
        return courseList;
    }

    // Pada metode ini di modul selanjutnya akan mengembalikan kelas POJO baru, gabungan antara course dengan module-nya.
    @Override
    public CourseEntity getCourseWithModules(final String courseId) {
        List<CourseResponse> courseResponses = remoteDataSource.getAllCourses();
        CourseEntity course = null;
        for (CourseResponse response : courseResponses) {
            if (response.getId().equals(courseId)) {
                course = new CourseEntity(response.getId(),
                        response.getTitle(),
                        response.getDescription(),
                        response.getDate(),
                        false,
                        response.getImagePath());
            }
        }
        return course;
    }

    @Override
    public ArrayList<ModuleEntity> getAllModulesByCourse(String courseId) {
        List<ModuleResponse> moduleResponses = remoteDataSource.getModules(courseId);
        ArrayList<ModuleEntity> moduleList = new ArrayList<>();
        for (ModuleResponse response : moduleResponses) {
            ModuleEntity course = new ModuleEntity(response.getModuleId(),
                    response.getCourseId(),
                    response.getTitle(),
                    response.getPosition(),
                    false);

            moduleList.add(course);
        }
        return moduleList;
    }


    @Override
    public ModuleEntity getContent(String courseId, String moduleId) {
        List<ModuleResponse> moduleResponses = remoteDataSource.getModules(courseId);
        ModuleEntity module = null;
        for (ModuleResponse response : moduleResponses) {
            if (response.getModuleId().equals(moduleId)) {
                module = new ModuleEntity(response.getModuleId(),
                        response.getCourseId(),
                        response.getTitle(),
                        response.getPosition(),
                        false);
                module.contentEntity = new ContentEntity(remoteDataSource.getContent(moduleId).getContent());
                break;
            }
        }
        return module;
    }
}

File: Project Academy/6. LiveData/kotlin/app/src/test/java/com/dicoding/academies/data/FakeAcademyRepository.kt
package com.dicoding.academies.data

import com.dicoding.academies.data.source.local.entity.ContentEntity
import com.dicoding.academies.data.source.local.entity.CourseEntity
import com.dicoding.academies.data.source.local.entity.ModuleEntity
import com.dicoding.academies.data.source.remote.RemoteDataSource
import java.util.*

class FakeAcademyRepository(private val remoteDataSource: RemoteDataSource) : AcademyDataSource {

    companion object {
        @Volatile
        private var instance: FakeAcademyRepository? = null

        fun getInstance(remoteData: RemoteDataSource): FakeAcademyRepository =
                instance ?: synchronized(this) {
                    instance ?: FakeAcademyRepository(remoteData)
                }
    }

    override fun getAllCourses(): ArrayList<CourseEntity> {
        val courseResponses = remoteDataSource.getAllCourses()
        val courseList = ArrayList<CourseEntity>()
        for (i in courseResponses.indices) {
            val response = courseResponses[i]
            val course = CourseEntity(response.id,
                    response.title,
                    response.description,
                    response.date,
                    false,
                    response.imagePath)

            courseList.add(course)
        }
        return courseList
    }


    override fun getBookmarkedCourses(): ArrayList<CourseEntity> {
        val courseList = ArrayList<CourseEntity>()
        val courses = remoteDataSource.getAllCourses()
        for (i in courses.indices) {
            val response = courses[i]
            val course = CourseEntity(response.id,
                    response.title,
                    response.description,
                    response.date,
                    false,
                    response.imagePath)
            courseList.add(course)
        }
        return courseList
    }

    // Pada metode ini di modul selanjutnya akan mengembalikan kelas POJO baru, gabungan antara course dengan module-nya.
    override fun getCourseWithModules(courseId: String?): CourseEntity {
        var course: CourseEntity? = null
        val courses = remoteDataSource.getAllCourses()
        for (i in courses.indices) {
            val response = courses[i]
            if (response.id == courseId) {
                course = CourseEntity(response.id,
                        response.title,
                        response.description,
                        response.date,
                        false,
                        response.imagePath)
            }
        }
        return course as CourseEntity
    }

    override fun getAllModulesByCourse(courseId: String?): ArrayList<ModuleEntity> {
        val moduleList = ArrayList<ModuleEntity>()
        val moduleResponses = remoteDataSource.getModules(courseId)
        for (i in moduleResponses.indices) {
            val response = moduleResponses[i]
            val course = ModuleEntity(response.moduleId,
                    response.courseId,
                    response.title,
                    response.position,
                    false)

            moduleList.add(course)
        }

        return moduleList
    }


    override fun getContent(courseId: String?, moduleId: String?): ModuleEntity {
        val moduleResponses = remoteDataSource.getModules(courseId)

        var module: ModuleEntity? = null
        for (i in moduleResponses.indices) {
            val moduleResponse = moduleResponses[i]

            val id = moduleResponse.moduleId

            if (id == moduleId) {
                module = ModuleEntity(id, moduleResponse.courseId, moduleResponse.title, moduleResponse.position, false)

                module.contentEntity = ContentEntity(remoteDataSource.getContent(moduleId).content)
                break
            }
        }

        return module as ModuleEntity
    }
}

